File: src/main/java/br/com/itau/governanca/risco/CalculadoraRisco.java
package br.com.itau.governanca.risco;

import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;

/**
 * Calculadora de Risco de Qualidade para aplicações bancárias
 * Implementa o modelo de risco para avaliar a qualidade de aplicações
 * baseado em quatro dimensões: bugs, performance, segurança e experiência
 */
public class CalculadoraRisco {
    
    // Pesos das dimensões
    private final Map<String, Double> pesosDimensoes;
    
    // Pesos dos critérios por dimensão
    private final Map<String, Map<String, Double>> pesosCriterios;
    
    // Limiares para classificação de risco
    private final Map<String, Map<String, Map<String, Object>>> limiares;
    
    /**
     * Construtor que inicializa os pesos e limiares do modelo de risco
     */
    public CalculadoraRisco() {
        // Inicializa os pesos das dimensões
        pesosDimensoes = new HashMap<>();
        pesosDimensoes.put("bugs", 0.25);
        pesosDimensoes.put("performance", 0.25);
        pesosDimensoes.put("seguranca", 0.30);
        pesosDimensoes.put("experiencia", 0.20);
        
        // Inicializa os pesos dos critérios
        pesosCriterios = new HashMap<>();
        
        // Pesos para Bugs
        Map<String, Double> pesosBugs = new HashMap<>();
        pesosBugs.put("densidade_defeitos", 0.25);
        pesosBugs.put("cobertura_testes", 0.20);
        pesosBugs.put("bugs_criticos", 0.30);
        pesosBugs.put("taxa_regressao", 0.15);
        pesosBugs.put("mtbf", 0.10);
        pesosCriterios.put("bugs", pesosBugs);
        
        // Pesos para Performance
        Map<String, Double> pesosPerformance = new HashMap<>();
        pesosPerformance.put("tempo_resposta", 0.25);
        pesosPerformance.put("percentil_95", 0.25);
        pesosPerformance.put("utilizacao_recursos", 0.15);
        pesosPerformance.put("escalabilidade", 0.15);
        pesosPerformance.put("tempo_inicializacao", 0.05);
        pesosPerformance.put("throughput", 0.15);
        pesosCriterios.put("performance", pesosPerformance);
        
        // Pesos para Segurança
        Map<String, Double> pesosSeguranca = new HashMap<>();
        pesosSeguranca.put("vulnerabilidades_criticas", 0.30);
        pesosSeguranca.put("vulnerabilidades_totais", 0.20);
        pesosSeguranca.put("owasp_top10", 0.20);
        pesosSeguranca.put("seguranca_dados", 0.15);
        pesosSeguranca.put("autenticacao_autorizacao", 0.10);
        pesosSeguranca.put("pentest", 0.05);
        pesosCriterios.put("seguranca", pesosSeguranca);
        
        // Pesos para Experiência
        Map<String, Double> pesosExperiencia = new HashMap<>();
        pesosExperiencia.put("satisfacao_usuario", 0.25);
        pesosExperiencia.put("taxa_erro_usuario", 0.20);
        pesosExperiencia.put("tempo_conclusao_tarefa", 0.15);
        pesosExperiencia.put("acessibilidade", 0.10);
        pesosExperiencia.put("taxa_abandono", 0.15);
        pesosExperiencia.put("adocao_funcionalidades", 0.15);
        pesosCriterios.put("experiencia", pesosExperiencia);
        
        // Inicializa os limiares
        limiares = new HashMap<>();
        
        // Limiares para Bugs
        Map<String, Map<String, Object>> limBugs = new HashMap<>();
        
        Map<String, Object> limDensidadeDefeitos = new HashMap<>();
        limDensidadeDefeitos.put("baixo", 2.0);
        limDensidadeDefeitos.put("medio", 5.0);
        limBugs.put("densidade_defeitos", limDensidadeDefeitos);
        
        Map<String, Object> limCoberturaTestes = new HashMap<>();
        limCoberturaTestes.put("baixo", 90.0);
        limCoberturaTestes.put("medio", 70.0);
        limCoberturaTestes.put("invertido", true);
        limBugs.put("cobertura_testes", limCoberturaTestes);
        
        Map<String, Object> limBugsCriticos = new HashMap<>();
        limBugsCriticos.put("baixo", 0.0);
        limBugsCriticos.put("medio", 2.0);
        limBugs.put("bugs_criticos", limBugsCriticos);
        
        Map<String, Object> limTaxaRegressao = new HashMap<>();
        limTaxaRegressao.put("baixo", 1.0);
        limTaxaRegressao.put("medio", 3.0);
        limBugs.put("taxa_regressao", limTaxaRegressao);
        
        Map<String, Object> limMTBF = new HashMap<>();
        limMTBF.put("baixo", 720.0);
        limMTBF.put("medio", 168.0);
        limMTBF.put("invertido", true);
        limBugs.put("mtbf", limMTBF);
        
        limiares.put("bugs", limBugs);
        
        // Limiares para Performance
        Map<String, Map<String, Object>> limPerformance = new HashMap<>();
        
        Map<String, Object> limTempoResposta = new HashMap<>();
        limTempoResposta.put("baixo", 300.0);
        limTempoResposta.put("medio", 800.0);
        limPerformance.put("tempo_resposta", limTempoResposta);
        
        Map<String, Object> limPercentil95 = new HashMap<>();
        limPercentil95.put("baixo", 800.0);
        limPercentil95.put("medio", 2000.0);
        limPerformance.put("percentil_95", limPercentil95);
        
        Map<String, Object> limUtilizacaoRecursos = new HashMap<>();
        limUtilizacaoRecursos.put("baixo", 60.0);
        limUtilizacaoRecursos.put("medio", 85.0);
        limPerformance.put("utilizacao_recursos", limUtilizacaoRecursos);
        
        Map<String, Object> limEscalabilidade = new HashMap<>();
        limEscalabilidade.put("baixo", 10.0);
        limEscalabilidade.put("medio", 30.0);
        limPerformance.put("escalabilidade", limEscalabilidade);
        
        Map<String, Object> limTempoInicializacao = new HashMap<>();
        limTempoInicializacao.put("baixo", 5.0);
        limTempoInicializacao.put("medio", 15.0);
        limPerformance.put("tempo_inicializacao", limTempoInicializacao);
        
        Map<String, Object> limThroughput = new HashMap<>();
        limThroughput.put("baixo", 100.0);
        limThroughput.put("medio", 50.0);
        limThroughput.put("invertido", true);
        limPerformance.put("throughput", limThroughput);
        
        limiares.put("performance", limPerformance);
        
        // Limiares para Segurança
        Map<String, Map<String, Object>> limSeguranca = new HashMap<>();
        
        Map<String, Object> limVulnCriticas = new HashMap<>();
        limVulnCriticas.put("baixo", 0.0);
        limVulnCriticas.put("medio", 1.0);
        limSeguranca.put("vulnerabilidades_criticas", limVulnCriticas);
        
        Map<String, Object> limVulnTotais = new HashMap<>();
        limVulnTotais.put("baixo", 5.0);
        limVulnTotais.put("medio", 15.0);
        limSeguranca.put("vulnerabilidades_totais", limVulnTotais);
        
        Map<String, Object> limOwaspTop10 = new HashMap<>();
        limOwaspTop10.put("baixo", 100.0);
        limOwaspTop10.put("medio", 90.0);
        limOwaspTop10.put("invertido", true);
        limSeguranca.put("owasp_top10", limOwaspTop10);
        
        Map<String, Object> limSegurancaDados = new HashMap<>();
        limSegurancaDados.put("baixo", 100.0);
        limSegurancaDados.put("medio", 95.0);
        limSegurancaDados.put("invertido", true);
        limSeguranca.put("seguranca_dados", limSegurancaDados);
        
        Map<String, Object> limAutenticacao = new HashMap<>();
        limAutenticacao.put("baixo", 2.0);
        limAutenticacao.put("medio", 1.0);
        limSeguranca.put("autenticacao_autorizacao", limAutenticacao);
        
        Map<String, Object> limPentest = new HashMap<>();
        limPentest.put("baixo", 2.0);
        limPentest.put("medio", 1.0);
        limSeguranca.put("pentest", limPentest);
        
        limiares.put("seguranca", limSeguranca);
        
        // Limiares para Experiência
        Map<String, Map<String, Object>> limExperiencia = new HashMap<>();
        
        Map<String, Object> limSatisfacaoUsuario = new HashMap<>();
        limSatisfacaoUsuario.put("baixo", 70.0);
        limSatisfacaoUsuario.put("medio", 40.0);
        limSatisfacaoUsuario.put("invertido", true);
        limExperiencia.put("satisfacao_usuario", limSatisfacaoUsuario);
        
        Map<String, Object> limTaxaErroUsuario = new HashMap<>();
        limTaxaErroUsuario.put("baixo", 2.0);
        limTaxaErroUsuario.put("medio", 5.0);
        limExperiencia.put("taxa_erro_usuario", limTaxaErroUsuario);
        
        Map<String, Object> limTempoConclusaoTarefa = new HashMap<>();
        limTempoConclusaoTarefa.put("baixo", 10.0);
        limTempoConclusaoTarefa.put("medio", 30.0);
        limExperiencia.put("tempo_conclusao_tarefa", limTempoConclusaoTarefa);
        
        Map<String, Object> limAcessibilidade = new HashMap<>();
        limAcessibilidade.put("baixo", 95.0);
        limAcessibilidade.put("medio", 80.0);
        limAcessibilidade.put("invertido", true);
        limExperiencia.put("acessibilidade", limAcessibilidade);
        
        Map<String, Object> limTaxaAbandono = new HashMap<>();
        limTaxaAbandono.put("baixo", 5.0);
        limTaxaAbandono.put("medio", 15.0);
        limExperiencia.put("taxa_abandono", limTaxaAbandono);
        
        Map<String, Object> limAdocaoFuncionalidades = new HashMap<>();
        limAdocaoFuncionalidades.put("baixo", 60.0);
        limAdocaoFuncionalidades.put("medio", 30.0);
        limAdocaoFuncionalidades.put("invertido", true);
        limExperiencia.put("adocao_funcionalidades", limAdocaoFuncionalidades);
        
        limiares.put("experiencia", limExperiencia);
    }
    
    /**
     * Classifica um critério como baixo (1), médio (2) ou alto (3) risco
     * @param dimensao A dimensão do critério (bugs, performance, seguranca, experiencia)
     * @param criterio O nome do critério
     * @param valor O valor a ser classificado
     * @return A classificação de risco (1=baixo, 2=médio, 3=alto)
     */
    public int classificarCriterio(String dimensao, String criterio, double valor) {
        Map<String, Object> limiar = limiares.get(dimensao).get(criterio);
        boolean invertido = limiar.containsKey("invertido") ? (boolean) limiar.get("invertido") : false;
        double limiarBaixo = (double) limiar.get("baixo");
        double limiarMedio = (double) limiar.get("medio");
        
        if (invertido) {
            if (valor >= limiarBaixo) {
                return 1; // Baixo risco
            } else if (valor >= limiarMedio) {
                return 2; // Médio risco
            } else {
                return 3; // Alto risco
            }
        } else {
            if (valor <= limiarBaixo) {
                return 1; // Baixo risco
            } else if (valor <= limiarMedio) {
                return 2; // Médio risco
            } else {
                return 3; // Alto risco
            }
        }
    }
    
    /**
     * Calcula a pontuação ponderada para uma dimensão
     * @param dimensao A dimensão (bugs, performance, seguranca, experiencia)
     * @param valores Um mapa com os valores dos critérios
     * @return A pontuação ponderada da dimensão
     */
    public double calcularPontuacaoDimensao(String dimensao, Map<String, Double> valores) {
        double pontuacaoTotal = 0;
        double pesoTotal = 0;
        
        Map<String, Double> pesosCriterioDimensao = pesosCriterios.get(dimensao);
        
        for (Map.Entry<String, Double> entry : valores.entrySet()) {
            String criterio = entry.getKey();
            double valor = entry.getValue();
            
            if (pesosCriterioDimensao.containsKey(criterio)) {
                double peso = pesosCriterioDimensao.get(criterio);
                int classificacao = classificarCriterio(dimensao, criterio, valor);
                pontuacaoTotal += classificacao * peso;
                pesoTotal += peso;
            }
        }
        
        // Garantir que temos pesos válidos
        if (pesoTotal == 0) {
            return 0;
        }
        
        return pontuacaoTotal / pesoTotal;
    }
    
    /**
     * Calcula o risco total com base nas pontuações das dimensões
     * @param dados Um mapa com os valores para todas as dimensões e critérios
     * @return Um mapa com o resultado da avaliação de risco
     */
    public Map<String, Object> calcularRiscoTotal(Map<String, Map<String, Double>> dados) {
        Map<String, Double> pontuacoesDimensoes = new HashMap<>();
        double riscoTotal = 0;
        
        for (Map.Entry<String, Map<String, Double>> entry : dados.entrySet()) {
            String dimensao = entry.getKey();
            Map<String, Double> criterios = entry.getValue();
            
            double pontuacao = calcularPontuacaoDimensao(dimensao, criterios);
            pontuacoesDimensoes.put(dimensao, pontuacao);
            riscoTotal += pontuacao * pesosDimensoes.get(dimensao);
        }
        
        // Classificação final
        String classificacao;
        if (riscoTotal <= 1.5) {
            classificacao = "VERDE (Baixo Risco)";
        } else if (riscoTotal <= 2.2) {
            classificacao = "AMARELO (Médio Risco)";
        } else {
            classificacao = "VERMELHO (Alto Risco)";
        }
        
        Map<String, Object> resultado = new HashMap<>();
        resultado.put("pontuacao_total", riscoTotal);
        resultado.put("classificacao", classificacao);
        resultado.put("pontuacoes_dimensoes", pontuacoesDimensoes);
        
        return resultado;
    }
    
    /**
     * Verifica se algum dos critérios de veto foi atingido
     * @param dados Um mapa com os valores para todas as dimensões e critérios
     * @return Uma lista com os critérios de veto atingidos
     */
    public List<String> verificarVetos(Map<String, Map<String, Double>> dados) {
        List<String> vetos = new ArrayList<>();
        
        // Veto 1: Qualquer vulnerabilidade crítica de segurança não mitigada
        if (dados.containsKey("seguranca") && 
            dados.get("seguranca").containsKey("vulnerabilidades_criticas") && 
            dados.get("seguranca").get("vulnerabilidades_criticas") > 0) {
            vetos.add("Vulnerabilidade crítica de segurança não mitigada");
        }
        
        // Veto 2: Tempo de resposta P95 > 3000ms em operações críticas
        if (dados.containsKey("performance") && 
            dados.get("performance").containsKey("percentil_95") && 
            dados.get("performance").get("percentil_95") > 3000) {
            vetos.add("Tempo de resposta P95 > 3000ms em operações críticas");
        }
        
        // Veto 3: Bugs que impeçam fluxo principal do cliente
        if (dados.containsKey("bugs") && 
            dados.get("bugs").containsKey("bugs_criticos") && 
            dados.get("bugs").get("bugs_criticos") > 0) {
            vetos.add("Bugs que impedem fluxo principal do cliente");
        }
        
        // Veto 4: Taxa de erro em produção >1% para operações críticas
        if (dados.containsKey("experiencia") && 
            dados.get("experiencia").containsKey("taxa_erro_usuario") && 
            dados.get("experiencia").get("taxa_erro_usuario") > 1) {
            vetos.add("Taxa de erro em produção >1% para operações críticas");
        }
        
        return vetos;
    }
    
    /**
     * Gera um relatório detalhado com o resultado da avaliação
     * @param nomeAplicacao Nome da aplicação
     * @param versao Versão da aplicação
     * @param dados Dados para avaliação
     * @return Um mapa com o relatório detalhado
     */
    public Map<String, Object> gerarRelatorio(String nomeAplicacao, String versao, 
                                             Map<String, Map<String, Double>> dados) {
        Map<String, Object> resultado = calcularRiscoTotal(dados);
        List<String> vetos = verificarVetos(dados);
        
        Map<String, Object> relatorio = new HashMap<>();
        relatorio.put("nome_aplicacao", nomeAplicacao);
        relatorio.put("versao", versao);
        relatorio.put("resultado", resultado);
        relatorio.put("vetos", vetos);
        
        boolean aprovado = vetos.isEmpty();
        boolean baixoRisco = resultado.get("classificacao").toString().startsWith("VERDE");
        
        String status;
        if (!aprovado) {
            status = "BLOQUEADO";
        } else if (baixoRisco) {
            status = "APROVADO";
        } else {
            status = "CONDICIONADO";
        }
        
        relatorio.put("status", status);
        relatorio.put("dados_detalhados", dados);
        
        return relatorio;
    }
}

File: documentacao/relatorios/calculadora_risco.py
#!/usr/bin/env python3
"""
Calculadora de Risco de Qualidade - Banco Itaú
Governança de Tecnologia

Este script calcula o nível de risco de uma aplicação com base nos critérios
estabelecidos no Modelo de Risco de Qualidade do Banco Itaú.
"""

import csv
import json
from datetime import datetime

class CalculadoraRisco:
    def __init__(self):
        # Definição dos pesos das dimensões
        self.pesos_dimensoes = {
            "bugs": 0.25,
            "performance": 0.25,
            "seguranca": 0.30,
            "experiencia": 0.20
        }
        
        # Definição dos pesos dos critérios por dimensão
        self.pesos_criterios = {
            "bugs": {
                "densidade_defeitos": 0.25,
                "cobertura_testes": 0.20,
                "bugs_criticos": 0.30,
                "taxa_regressao": 0.15,
                "mtbf": 0.10
            },
            "performance": {
                "tempo_resposta": 0.25,
                "percentil_95": 0.25,
                "utilizacao_recursos": 0.15,
                "escalabilidade": 0.15,
                "tempo_inicializacao": 0.05,
                "throughput": 0.15
            },
            "seguranca": {
                "vulnerabilidades_criticas": 0.30,
                "vulnerabilidades_totais": 0.20,
                "owasp_top10": 0.20,
                "seguranca_dados": 0.15,
                "autenticacao_autorizacao": 0.10,
                "pentest": 0.05
            },
            "experiencia": {
                "satisfacao_usuario": 0.25,
                "taxa_erro_usuario": 0.20,
                "tempo_conclusao_tarefa": 0.15,
                "acessibilidade": 0.10,
                "taxa_abandono": 0.15,
                "adocao_funcionalidades": 0.15
            }
        }
        
        # Limiares para classificação de risco
        self.limiares = {
            "bugs": {
                "densidade_defeitos": {"baixo": 2, "medio": 5},
                "cobertura_testes": {"baixo": 90, "medio": 70, "invertido": True},
                "bugs_criticos": {"baixo": 0, "medio": 2},
                "taxa_regressao": {"baixo": 1, "medio": 3},
                "mtbf": {"baixo": 720, "medio": 168, "invertido": True}
            },
            "performance": {
                "tempo_resposta": {"baixo": 300, "medio": 800},
                "percentil_95": {"baixo": 800, "medio": 2000},
                "utilizacao_recursos": {"baixo": 60, "medio": 85},
                "escalabilidade": {"baixo": 10, "medio": 30},
                "tempo_inicializacao": {"baixo": 5, "medio": 15},
                "throughput": {"baixo": 100, "medio": 50, "invertido": True}
            },
            "seguranca": {
                "vulnerabilidades_criticas": {"baixo": 0, "medio": 1},
                "vulnerabilidades_totais": {"baixo": 5, "medio": 15},
                "owasp_top10": {"baixo": 100, "medio": 90, "invertido": True},
                "seguranca_dados": {"baixo": 100, "medio": 95, "invertido": True},
                "autenticacao_autorizacao": {"baixo": 2, "medio": 1},  # 2=Completo, 1=Parcial, 0=Mínimo
                "pentest": {"baixo": 2, "medio": 1}  # 2=Sem falhas críticas, 1=Médio risco, 0=Alto risco
            },
            "experiencia": {
                "satisfacao_usuario": {"baixo": 70, "medio": 40, "invertido": True},
                "taxa_erro_usuario": {"baixo": 2, "medio": 5},
                "tempo_conclusao_tarefa": {"baixo": 10, "medio": 30},
                "acessibilidade": {"baixo": 95, "medio": 80, "invertido": True},
                "taxa_abandono": {"baixo": 5, "medio": 15},
                "adocao_funcionalidades": {"baixo": 60, "medio": 30, "invertido": True}
            }
        }
        
    def classificar_criterio(self, dimensao, criterio, valor):
        """Classifica um critério como baixo (1), médio (2) ou alto (3) risco."""
        limiar = self.limiares[dimensao][criterio]
        invertido = limiar.get("invertido", False)
        
        if invertido:
            if valor >= limiar["baixo"]:
                return 1  # Baixo risco
            elif valor >= limiar["medio"]:
                return 2  # Médio risco
            else:
                return 3  # Alto risco
        else:
            if valor <= limiar["baixo"]:
                return 1  # Baixo risco
            elif valor <= limiar["medio"]:
                return 2  # Médio risco
            else:
                return 3  # Alto risco
    
    def calcular_pontuacao_dimensao(self, dimensao, valores):
        """Calcula a pontuação ponderada para uma dimensão."""
        pontuacao_total = 0
        peso_total = 0
        
        for criterio, valor in valores.items():
            if criterio in self.pesos_criterios[dimensao]:
                peso = self.pesos_criterios[dimensao][criterio]
                classificacao = self.classificar_criterio(dimensao, criterio, valor)
                pontuacao_total += classificacao * peso
                peso_total += peso
        
        # Garantir que temos pesos válidos
        if peso_total == 0:
            return 0
            
        return pontuacao_total / peso_total
    
    def calcular_risco_total(self, dados):
        """Calcula o risco total com base nas pontuações das dimensões."""
        pontuacoes_dimensoes = {}
        risco_total = 0
        
        for dimensao, criterios in dados.items():
            pontuacao = self.calcular_pontuacao_dimensao(dimensao, criterios)
            pontuacoes_dimensoes[dimensao] = pontuacao
            risco_total += pontuacao * self.pesos_dimensoes[dimensao]
        
        # Classificação final
        if risco_total <= 1.5:
            classificacao = "VERDE (Baixo Risco)"
        elif risco_total <= 2.2:
            classificacao = "AMARELO (Médio Risco)"
        else:
            classificacao = "VERMELHO (Alto Risco)"
        
        return {
            "pontuacao_total": risco_total,
            "classificacao": classificacao,
            "pontuacoes_dimensoes": pontuacoes_dimensoes
        }
    
    def verificar_vetos(self, dados):
        """Verifica se algum dos critérios de veto foi atingido."""
        vetos = []
        
        # Veto 1: Qualquer vulnerabilidade crítica de segurança não mitigada
        if dados.get("seguranca", {}).get("vulnerabilidades_criticas", 0) > 0:
            vetos.append("Vulnerabilidade crítica de segurança não mitigada")
        
        # Veto 2: Tempo de resposta P95 > 3000ms em operações críticas
        if dados.get("performance", {}).get("percentil_95", 0) > 3000:
            vetos.append("Tempo de resposta P95 > 3000ms em operações críticas")
        
        # Veto 3: Bugs que impeçam fluxo principal do cliente
        if dados.get("bugs", {}).get("bugs_criticos", 0) > 0:
            vetos.append("Bugs que impedem fluxo principal do cliente")
        
        # Veto 4: Taxa de erro em produção >1% para operações críticas
        if dados.get("experiencia", {}).get("taxa_erro_usuario", 0) > 1:
            vetos.append("Taxa de erro em produção >1% para operações críticas")
        
        # Veto 5: não podemos verificar automaticamente
        
        return vetos
    
    def gerar_relatorio(self, nome_aplicacao, versao, dados):
        """Gera um relatório detalhado com o resultado da avaliação."""
        resultado = self.calcular_risco_total(dados)
        vetos = self.verificar_vetos(dados)
        
        relatorio = {
            "nome_aplicacao": nome_aplicacao,
            "versao": versao,
            "data_avaliacao": datetime.now().strftime("%d/%m/%Y"),
            "resultado": resultado,
            "vetos": vetos,
            "status": "BLOQUEADO" if vetos else "APROVADO" if resultado["classificacao"].startswith("VERDE") else "CONDICIONADO",
            "dados_detalhados": dados
        }
        
        return relatorio
    
    def carregar_dados_csv(self, arquivo_csv):
        """Carrega dados de um arquivo CSV formatado corretamente."""
        dados = {
            "bugs": {},
            "performance": {},
            "seguranca": {},
            "experiencia": {}
        }
        
        mapeamento_dimensoes = {
            "Bugs": "bugs",
            "Performance": "performance",
            "Segurança": "seguranca",
            "Experiência": "experiencia"
        }
        
        mapeamento_criterios = {
            "Densidade de defeitos": "densidade_defeitos",
            "Cobertura de testes": "cobertura_testes",
            "Bugs críticos pendentes": "bugs_criticos",
            "Taxa de regressão": "taxa_regressao",
            "MTBF": "mtbf",
            "Tempo de resposta": "tempo_resposta",
            "Percentil 95 de tempo de resposta": "percentil_95",
            "Utilização de recursos": "utilizacao_recursos",
            "Escalabilidade": "escalabilidade",
            "Tempo de inicialização": "tempo_inicializacao",
            "Throughput": "throughput",
            "Vulnerabilidades críticas": "vulnerabilidades_criticas",
            "Vulnerabilidades totais": "vulnerabilidades_totais",
            "OWASP Top 10": "owasp_top10",
            "Segurança de dados": "seguranca_dados",
            "Autenticação/Autorização": "autenticacao_autorizacao",
            "Pentest": "pentest",
            "Satisfação do usuário": "satisfacao_usuario",
            "Taxa de erro do usuário": "taxa_erro_usuario",
            "Tempo de conclusão de tarefa": "tempo_conclusao_tarefa",
            "Acessibilidade": "acessibilidade",
            "Taxa de abandono": "taxa_abandono",
            "Adoção de funcionalidades": "adocao_funcionalidades"
        }
        
        try:
            with open(arquivo_csv, 'r', encoding='utf-8') as f:
                leitor = csv.DictReader(f)
                for linha in leitor:
                    if 'Dimensão' in linha and 'Critério' in linha and 'Valor Atual' in linha:
                        dimensao = mapeamento_dimensoes.get(linha['Dimensão'])
                        criterio = mapeamento_criterios.get(linha['Critério'])
                        valor = linha['Valor Atual']
                        
                        if dimensao and criterio and valor and valor.strip():
                            try:
                                dados[dimensao][criterio] = float(valor)
                            except ValueError:
                                print(f"Aviso: Valor inválido para {criterio}: {valor}")
            
            return dados
        except Exception as e:
            print(f"Erro ao carregar CSV: {e}")
            return None
    
    def salvar_relatorio(self, relatorio, arquivo_saida):
        """Salva o relatório em formato JSON."""
        try:
            with open(arquivo_saida, 'w', encoding='utf-8') as f:
                json.dump(relatorio, f, indent=2, ensure_ascii=False)
            print(f"Relatório salvo em {arquivo_saida}")
            return True
        except Exception as e:
            print(f"Erro ao salvar relatório: {e}")
            return False

if __name__ == "__main__":
    import sys
    
    if len(sys.argv) < 4:
        print("Uso: python calculadora_risco.py <nome_aplicacao> <versao> <arquivo_csv> [arquivo_saida]")
        sys.exit(1)
    
    nome_aplicacao = sys.argv[1]
    versao = sys.argv[2]
    arquivo_csv = sys.argv[3]
    arquivo_saida = sys.argv[4] if len(sys.argv) > 4 else f"relatorio_{nome_aplicacao.replace(' ', '_')}_{versao}.json"
    
    calc = CalculadoraRisco()
    dados = calc.carregar_dados_csv(arquivo_csv)
    
    if dados:
        relatorio = calc.gerar_relatorio(nome_aplicacao, versao, dados)
        
        # Imprimir resumo na tela
        print(f"\nRelatório de Risco - {nome_aplicacao} v{versao}")
        print("-" * 50)
        print(f"Classificação: {relatorio['resultado']['classificacao']}")
        print(f"Pontuação: {relatorio['resultado']['pontuacao_total']:.2f}")
        print(f"Status: {relatorio['status']}")
        
        if relatorio['vetos']:
            print("\nCritérios de veto atingidos:")
            for veto in relatorio['vetos']:
                print(f"- {veto}")
        
        print("\nPontuações por dimensão:")
        for dim, pontuacao in relatorio['resultado']['pontuacoes_dimensoes'].items():
            nome_dim = {"bugs": "Bugs", "performance": "Performance", 
                       "seguranca": "Segurança", "experiencia": "Experiência"}
            print(f"- {nome_dim[dim]}: {pontuacao:.2f}")
        
        # Salvar relatório completo
        calc.salvar_relatorio(relatorio, arquivo_saida)
    else:
        print("Não foi possível processar os dados de entrada.")
        sys.exit(1)